pallets/xcm-benchmarks/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pallet that serves no other purpose than benchmarking raw messages [`Xcm`].
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use frame_benchmarking::{account, BenchmarkError};
+use sp_std::prelude::*;
+use xcm::latest::prelude::*;
+use xcm_executor::traits::Convert;
+
+pub mod fungible;
+pub mod generic;
+
+#[cfg(test)]
+mod mock;
+
+/// A base trait for all individual pallets
+pub trait Config: frame_system::Config {
+	/// The XCM configurations.
+	///
+	/// These might affect the execution of XCM messages, such as defining how the
+	/// `TransactAsset` is implemented.
+	type XcmConfig: xcm_executor::Config;
+
+	/// A converter between a multi-location to a sovereign account.
+	type AccountIdConverter: Convert<MultiLocation, Self::AccountId>;
+
+	/// Does any necessary setup to create a valid destination for XCM messages.
+	/// Returns that destination's multi-location to be used in benchmarks.
+	fn valid_destination() -> Result<MultiLocation, BenchmarkError>;
+
+	/// Worst case scenario for a holding account in this runtime.
+	fn worst_case_holding() -> MultiAssets;
+}
+
+const SEED: u32 = 0;
+
+/// The XCM executor to use for doing stuff.
+pub type ExecutorOf<T> = xcm_executor::XcmExecutor<<T as Config>::XcmConfig>;
+/// The overarching call type.
+pub type OverArchingCallOf<T> = <T as frame_system::Config>::Call;
+/// The asset transactor of our executor
+pub type AssetTransactorOf<T> = <<T as Config>::XcmConfig as xcm_executor::Config>::AssetTransactor;
+/// The call type of executor's config. Should eventually resolve to the same overarching call type.
+pub type XcmCallOf<T> = <<T as Config>::XcmConfig as xcm_executor::Config>::Call;
+
+pub fn mock_worst_case_holding() -> MultiAssets {
+	const HOLDING_FUNGIBLES: u32 = 99;
+	const HOLDING_NON_FUNGIBLES: u32 = 99;
+	let fungibles_amount: u128 = 100;
+	(0..HOLDING_FUNGIBLES)
+		.map(|i| {
+			MultiAsset {
+				id: Concrete(GeneralIndex(i as u128).into()),
+				fun: Fungible(fungibles_amount * i as u128),
+			}
+			.into()
+		})
+		.chain(core::iter::once(MultiAsset { id: Concrete(Here.into()), fun: Fungible(u128::MAX) }))
+		.chain((0..HOLDING_NON_FUNGIBLES).map(|i| MultiAsset {
+			id: Concrete(GeneralIndex(i as u128).into()),
+			fun: NonFungible(asset_instance_from(i)),
+		}))
+		.collect::<Vec<_>>()
+		.into()
+}
+
+pub fn asset_instance_from(x: u32) -> AssetInstance {
+	let bytes = x.encode();
+	let mut instance = [0u8; 4];
+	instance.copy_from_slice(&bytes);
+	AssetInstance::Array4(instance)
+}
+
+pub fn new_executor<T: Config>(origin: MultiLocation) -> ExecutorOf<T> {
+	ExecutorOf::<T>::new(origin)
+}
+
+/// Build a multi-location from an account id.
+fn account_id_junction<T: frame_system::Config>(index: u32) -> Junction {
```

Comment:
I wonder if any of this code can be reused from polkadot?

Reply:
At the moment we're experimenting with having a completely detached version of an xcm-benchmark pallet in cumulus to see how far it's going to diverge with a possibility of merging it back into the polkadot going forward and making it more generic. There's a ticket to revisit all this after this PR has been merged and released.